src/terms.jl

```diff
@@ -0,0 +1,524 @@
+# for a single row, some will be scalars, others possibly vectors.  for a whole
+# table, some will be vectors, possibly some matrices
+function kron_insideout(op::Function, args...)
+    args = (reshape(a, ones(Int, i-1)..., :) for (i,a) in enumerate(args))
+    vec(broadcast(op, args...))
+end
+
+function row_kron_insideout(op::Function, args...)
+    rows = size(args[1], 1)
+    args = (reshape(a, size(a,1), ones(Int, i-1)..., :) for (i,a) in enumerate(args))
+    reshape(broadcast(op, args...), rows, :)
+end
+
+# two options here: either special-case ColumnTable (named tuple of vectors)
+# vs. vanilla NamedTuple, or reshape and use normal broadcasting
+model_cols(t::InteractionTerm, d::NamedTuple) =
+    kron_insideout(*, (model_cols(term, d) for term in t.terms)...)
+
+function model_cols(t::InteractionTerm, d::ColumnTable)
+    row_kron_insideout(*, (model_cols(term, d) for term in t.terms)...)
+end
+
+model_cols(t::InterceptTerm{true}, d::NamedTuple) = ones(size(first(d)))
+model_cols(t::InterceptTerm{false}, d) = Matrix{Float64}(undef, size(first(d),1), 0)
+
+model_cols(t::FormulaTerm, d::NamedTuple) = (model_cols(t.lhs,d), model_cols(t.rhs, d))
+
+function model_cols(t::MatrixTerm, d::ColumnTable)
+    mat = reduce(hcat, (model_cols(tt, d) for tt in t.terms))
+    reshape(mat, size(mat, 1), :)
+end
+
+model_cols(t::MatrixTerm, d::NamedTuple) =
+    reduce(vcat, (model_cols(tt, d) for tt in t.terms))
+
+vectorize(x::Tuple) = collect(x)
+vectorize(x::AbstractVector) = x
+vectorize(x) = [x]
+
+"""
+    coefnames(term::AbstractTerm)
+
+Return the name[s] of columns generated by a term.  Return value is either a
+`String` or an iterable of `String`s.
+"""
+StatsBase.coefnames(t::FormulaTerm) = (coefnames(t.lhs), coefnames(t.rhs))
+StatsBase.coefnames(::InterceptTerm{H}) where {H} = H ? "(Intercept)" : []
+StatsBase.coefnames(t::ContinuousTerm) = string(t.sym)
+StatsBase.coefnames(t::CategoricalTerm) = 
+    ["$(t.sym): $name" for name in t.contrasts.termnames]
+StatsBase.coefnames(t::FunctionTerm) = string(t.exorig)
+StatsBase.coefnames(ts::TupleTerm) = reduce(vcat, coefnames.(ts))
+StatsBase.coefnames(t::MatrixTerm) = coefnames(t.terms)
+StatsBase.coefnames(t::InteractionTerm) =
+    kron_insideout((args...) -> join(args, " & "), vectorize.(coefnames.(t.terms))...)
+
+################################################################################
+# old Terms features:
+
+hasintercept(t::AbstractTerm) = InterceptTerm{true}() ∈ terms(t) || ConstantTerm(1) ∈ terms(t)
+hasnointercept(t::AbstractTerm) =
+    InterceptTerm{false}() ∈ terms(t) ||
+    ConstantTerm(0) ∈ terms(t) ||
+    ConstantTerm(-1) ∈ terms(t)
+
+hasresponse(t) = false
+hasresponse(t::FormulaTerm{LHS}) where {LHS} = LHS !== nothing
+
+# convenience converters
+"""
+    term(xs...)
+
+Wrap arguments in an appropriate `AbstractTerm` type: `Symbol`s become `Term`s,
+and `Number`s become `ConstantTerm`s.  Any `AbstractTerm`s are unchanged.
```

Comment:
Yeah I think the multi-arg form was something I did just because I was sick of writing `term.((1, :a, :b))` etc.  The broadcasting form is more explicit for sure but then again the point of this function is convenience (so you don't have to write `[ConstantTerm(1), Term(:a), Term(:b)]` etc.) so I'm not sure what the priority is...

Reply:
`term.((1, :a, :b))` doesn't sound too bad to me, but well... (I must say in general I always tend to limit the public API the the bare minimum, and see later whether it's too inconvenient. Maybe that's extreme...)

Reply:
Agreed; there's only a few places where this was used so I just removed the auto-broadcasting method. (also makes it clearer when there's not a `term` method defined for a particular type...)